### PR TITLE
Autofill Manual Programming Question Score

### PR DESCRIPTION
**Pull request by bivanalhar**
In the manual programming question and MCQ/MRQ question, upon checking the submission, all the scores will be initially empty. However, it's actually quite obvious when all the test-cases (for programming) are passed or when all the options chosen (for MCQ/MRQ) are correct, the score shall be the maximum score possible.

Therefore, upon evaluating the question, we have the grade box be autofilled with maximum score if all test cases are considered correct. However, those autofilled scores should not be directly saved into the system, and hence the warning sign will appear besides each of the grade, to warn users to save the grade by clicking the "Save Grade" button below.

Screenshot 2023-09-04 at 9 51 18 AM

### CODE REVIEW

**Comment by benleong** [comment]
These scores should be automatically generated at the frontend, but they should not be auto-saved. We should only autosave human inputs.

**Comment by bivanalhar** [comment]
yes Prof. Those scores are only autofilled in the FE. You still need to press the button "Save Grade" at the button to save the grades into the system

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditStepForm.jsx [hunk 1]
```diff
@@ -293,7 +294,16 @@ const SubmissionEditStepForm = (props) => {
     const editable = !attempting && graderView;
     const visible = editable || published;
 
-    return visible ? <QuestionGrade editable={editable} id={id} /> : null;
+    return visible ? (
+      <QuestionGrade
+        editable={editable}
+        explanations={explanations}
+        gradeIsSaved={gradeIsSaved}
+        gradeIsUnsavedMessage={intl.formatMessage(translations.gradeUnsaved)}
```
Comment (purfectliterature) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditStepForm.jsx:302]: Any reason why this message has to be configurable? Looks like it's pretty standard. `QuestionGrade` could import it itself from `translations.ts`.

**Review by purfectliterature — APPROVED** [review]
@bivanalhar Some observations about your approach:
- Your "unsaved state" (dirty state) indicator doesn't work because once the field is changed, it will never go back to non-unsaved state (i.e., it will always be marked as unsaved). This state is more aptly known as "touched state".
- `QuestionGrade` is now coupled with the programming and MCQ answers logic because it tried to determine if it's all correct before updating the field with the maximum grade.
- Also, you updated the field to the maximum grade by dispatching the update action _during the page render_. This not only meant there will be extra N renders (where N is the number of correct answers), but also potentially mess with the render lifecycle.
- Your approach pre-fills even with programming questions that fail some evaluation tests.

Since you are on MC, here I have shown another approach that minimises the changes of front-facing components; only `QuestionGrade` changed, and that's only to add the dirty and pre-filled indicators. The logic for pre-filling the maximum grade doesn't happen in the component (i.e., after render), rather happens in Redux, right after we receive the data from the server and during the processing.

This is how it looks now. I also added an indicator that shows if a grade has been pre-filled, and added a yellow border for increased visual clarity.

>[!NOTE]
We should refrain from using red(dish) colours unless we're sure it's a destructive or erroneous states. This is to make sure users don't tune out the strength of red(dish) colours in our app.

image

Feel free to refer to the commit history and various comments I left for you. **The comments I left are for reference only.**

I've given my LGTM. @ekowidianto Do you want to take one last look before we merge this in?

**Comment by bivanalhar** [comment]
Hi @purfectliterature thanks a lot for your comments and also for providing the better alternatives regarding the code style and also the reasoning behind it. I learnt new things from your proposed code, and I agreed with it

Just maybe one small change from me regarding the UI.
Screenshot 2023-09-07 at 9 22 07 AM

I think in this case, it's much better to put those indicators of being `Pre-filled` and `Unsaved` just beside the grade. This is because one of the requirements is to put the `{grade}/{maxGrade}` beside the text Grade itself due to the potential of the very wide screen being used. It will be easier for the grader to see whether the grade filled is actually from the pre-filled and thus unsaved if it's put just right to the grade itself.

What do you think? I'm pushing it now, if you think the new one is not okay, feel free to revert it back :)

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditStepForm.jsx [hunk 2]
```diff
@@ -112,6 +112,7 @@ const SubmissionEditStepForm = (props) => {
   const initialStep = Math.min(maxInitialStep, Math.max(0, step || 0));
 
   const [submitConfirmation, setSubmitConfirmation] = useState(false);
+  const [gradeIsSaved, setGradeIsSaved] = useState(true);
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditStepForm.jsx]: yes you're right. It's indeed what I also think when I'm doing this code. I was planning to do the dirty mode, but failed to find the way how to, and hence this is the one I can think of.

Now I looked into your proposed code, I learnt something new, thanks for that :)

**Review by ekowidianto — APPROVED** [review]
Good effort @bivanalhar and thanks for the review @purfectliterature - good comments. LGTM

File: app/views/course/assessment/answer/programming/_programming.json.jbuilder [hunk 3]
```diff
@@ -99,7 +101,9 @@ json.explanation do
       json.failureType 'private_test'
     end
 
-    json.correct attempt&.auto_grading && attempt&.correct
+    passed_evaluation_tests = failed_test_cases_by_type['evaluation_test'].blank?
+
+    json.correct attempt&.auto_grading && attempt&.correct && (can_grade && passed_evaluation_tests)
```
Comment (ekowidianto) [inline app/views/course/assessment/answer/programming/_programming.json.jbuilder:106]: Actually, I realise if a student accesses the page, then `json.correct` becomes `false`? @purfectliterature

Comment (purfectliterature) [reply app/views/course/assessment/answer/programming/_programming.json.jbuilder:106]: Right, I meant to add `can_grade ? passed_evaluation_tests : true` to not change anything for students, and only consider evaluation tests for graders.

I will make another PR for this.

**Pull request merged**